source/mqtt3_to_mqtt5_adapter.c

```diff
@@ -0,0 +1,179 @@
+    struct aws_allocator *allocator;
+
+    struct aws_mqtt_client_connection base;
+
+    struct aws_mqtt5_client *client;
+    struct aws_mqtt5_listener *listener;
+    struct aws_event_loop *loop;
+
+    struct aws_mutex state_lock;
+    enum aws_mqtt5_adapter_state state;
+};
+
+static void s_aws_mqtt5_client_connection_event_callback_adapter(const struct aws_mqtt5_client_lifecycle_event *event) {
+    (void)event;
+}
+
+static bool s_aws_mqtt5_listener_publish_received_adapter(
+    const struct aws_mqtt5_packet_publish_view *publish,
+    void *user_data) {
+    (void)publish;
+    (void)user_data;
+
+    return false;
+}
+
+static void s_disable_adapter(struct aws_mqtt_client_connection_5_impl *adapter) {
+    /*
+     * The lock is held during callbacks to prevent invoking into something that is in the process of
+     * destruction.  In general this isn't a performance worry since callbacks are invoked from a single
+     * thread: the event loop that the client and adapter are seated on.
+     *
+     * But since we don't have recursive mutexes on all platforms, we need to be careful about the disable
+     * API since if we naively always locked, then an adapter release inside a callback would deadlock.
+     *
+     * On the surface, it seems reasonable that if we're in the event loop thread we could just skip
+     * locking entirely (because we've already locked it at the start of the callback).  Unfortunately, this isn't safe
+     * because we don't actually truly know we're in our mqtt5 client's callback; we could be in some other
+     * client/connection's callback that happens to be seated on the same event loop.  And while it's true that because
+     * of the thread seating, nothing will be interfering with our shared state manipulation, there's one final
+     * consideration which forces us to *try* to lock:
+     *
+     * Dependent on the memory model of the CPU architecture, changes to shared state, even if "safe" from data
+     * races across threads, may not become visible to other cores on the same CPU unless some kind of synchronization
+     * primitive (memory barrier) is invoked.  So in this extremely unlikely case, we use try-lock to guarantee that a
+     * synchronization primitive is invoked when disable is coming through a callback from something else on the same
+     * event loop.
+     *
+     * In the case that we're in our mqtt5 client's callback, the lock is already held, try fails, and the unlock at
+     * the end of the callback will suffice for cache flush and synchronization.
+     *
+     * In the case that we're in something else's callback on the same thread, the try succeeds and its followup
+     * unlock here will suffice for cache flush and synchronization.
+     */
+    if (aws_event_loop_thread_is_callers_thread(adapter->loop)) {
+        bool lock_succeeded = aws_mutex_try_lock(&adapter->state_lock) == AWS_OP_SUCCESS;
+        adapter->state = AWS_MQTT5_AS_DISABLED;
```

Comment:
Question: Does this only need to be set if the lock succeeded?

Reply:
No it must be unconditionally set.

Reply:
Reading through the comment again, I think I understand now. 👍